### DISCUSSION TITLE
Usage of ipytone.Multiply

### DISCUSSION THREAD

**Discussion by kumolonimbus** [geigr/ipytone#119] [Q&A]
Hi! I've been trying to use ipytone to replicate a simple Max patch that multiplies an input signal by a second control signal. In Max, this is done with a [signal multiplier-operator](https://docs.cycling74.com/max8/refpages/times~). It seems like a similar result could be achieved with `ipytone.Multiply`, but I can't figure out how to pass the input and control signals to it, or how to hear the output.

I tried the following, where the input signal is `player`, the control signal is `osc` and an instance of `ipytone.Multiply` is `mult`:

```
player.connect(mult)
osc.connect(mult)
```

Checking the audio graph's connections gives me the following:

```
[(Player(name='bird_player'),
  Destination(name='main output', volume=Param(value=0.0, units='decibels'), mute=False),
  0,
  0),
 (Oscillator(name='phasor'),
  Destination(name='main output', volume=Param(value=0.0, units='decibels'), mute=False),
  0,
  0),
 (Player(name='bird_player'),
  Multiply(factor=Param(value=1.0, units='number')),
  0,
  0),
 (Oscillator(name='phasor'),
  Multiply(factor=Param(value=1.0, units='number')),
  0,
  0)]
```

If I try to run `osc.start()` and `player.start()`, I hear both signals in parallel (i.e., not being multiplied and played as a single signal). However, trying to run `mult.start()` gives an error message "'Multiply' object has no attribute 'start'". Any advice on how to get this working would be greatly appreciated!

**Comment by benbovy** [comment]
Hi @kumolonimbus, I think what you want is:

```python
player.connect(mult)
osc.connect(mult.factor)

# assuming that you want to hear the output of mult directly
mult.to_destination()

player.start()
osc.start()
```

[Multiply.factor](https://ipytone.readthedocs.io/en/latest/_api_generated/ipytone.Multiply.html#ipytone.Multiply.factor) is the control parameter and is an instance of [Param](https://ipytone.readthedocs.io/en/latest/_api_generated/ipytone.Param.html#ipytone.Param) to which you can connect another signal.

Connecting two signals to `Multiply` will just pass them through with a multiplication factor of 1.

I think it would be worth adding a section about audio signals in ipytone's documentation! In general ipytone closely follows the Tonejs API so checking [Tonejs API documentation](https://tonejs.github.io/docs/14.7.77/index.html) is useful for ipytone too.

> trying to run mult.start() gives an error message "'Multiply' object has no attribute 'start'"

`Multiply` is not a source node, it doesn't have a `start()` method.

**Comment by kumolonimbus** [comment]
Ah, the logic is a lot clearer now -- thanks for the suggestion @benbovy! I gave this a try, and the updated audio graph is as follows:

```
[(Multiply(factor=Param(value=1.0, units='number')),
  Destination(name='main output', volume=Param(value=0.0, units='decibels'), mute=False),
  0,
  0),
 (Player(name='bird_player'),
  Multiply(factor=Param(value=1.0, units='number')),
  0,
  0),
 (Oscillator(name='phasor'), Param(value=1.0, units='number'), 0, 0)]
```

This _looks_ correct to me: Player is connected to Multiply, the oscillator is connected to the multiplication factor, then Multiply is connected to the speaker. However after calling `player.start()` and `osc.start()`, I seem to only be hearing the output of the player. My guess is that the oscillator is not updating the [Multiply.factor](https://ipytone.readthedocs.io/en/latest/_api_generated/ipytone.Multiply.html#ipytone.Multiply.factor) values. The audio graph shows five nodes:

```
[Oscillator(name='phasor'),
 Param(value=1.0, units='number'),
 Player(name='bird_player'),
 Multiply(factor=Param(value=1.0, units='number')),
 Destination(name='main output', volume=Param(value=0.0, units='decibels'), mute=False)]
```

Wondering if the repetition of Param is causing this somehow?

**Reply by benbovy** [reply]
Hmm that's weird... I've tried with the following setup, which displays an `Output` widget as cell output:

```python
import ipytone
import ipywidgets

output = ipywidgets.Output()

def print_change(change):
    with output:
        print(change["new"])

sig = ipytone.Signal(value=2)
osc = ipytone.Oscillator()
mult = ipytone.Multiply()

sig.connect(mult)
osc.connect(mult.factor)

meter = ipytone.DCMeter()
mult.connect(meter)

meter.schedule_observe(print_change, update_interval=0.5, name="value")

output
```

Then, when I do `osc.start()` I can see from the `Output` widget the value of mult being modulated:

```
2.0
0.9732755422592163
1.094151496887207
3.991612672805786
2.6153810024261475
3.2828798294067383
3.7783803939819336
1.8273468017578125
0.40552330017089844
3.8710522651672363
...
```

Note that in Tone.js (and so in ipytone), the output of `Multiply` is a positive number. So probably you would need to convert the output of `Multiply` back into the [-1, 1] audio range (you could use [WaveShaper](https://ipytone.readthedocs.io/en/latest/_api_generated/ipytone.WaveShaper.html) for that).

I've not seen the Max patch you're trying to reproduce, and actually I've not used much Max (or PureData), so might be missing bits. Are you sure that multiplying the signal from a player by an oscillator is what you want? The frequency of oscillators are very high. Wouldn't be a `LFO` instead?

**Reply by benbovy** [reply]
Sorry https://github.com/benbovy/ipytone/discussions/119#discussioncomment-6151994.

**Comment by benbovy** [comment]
> So probably you would need to convert the output of Multiply back into the [-1, 1] audio range (you could use [WaveShaper](https://ipytone.readthedocs.io/en/latest/_api_generated/ipytone.WaveShaper.html) for that).

EDIT: I don't think the `WaveShaper` is going to help here, but you could probably include other signal nodes between `osc` and `mult` or elsewhere in the graph to ensure that the signal stays within the [-1, 1] audio range. Unfortunately, not all Tonejs signal nodes are available in ipytone currently but this should be reasonably straightforward to add the missing nodes (#82).

**Comment by kumolonimbus** [comment]
The `DCMeter` was great for troubleshooting! I managed to get it to work, though honestly it's not super clear to me why the current version works and the previous version didn't. The main changes I made were to increase the volume of the oscillator and work in JupyterLab instead of VSCode.

My final audio graph:
```
(Player(), Multiply(factor=Param(value=1.0, units='number')), 0, 0),
 (Oscillator(name='phasor'), Param(value=1.0, units='number'), 0, 0),
 (Multiply(factor=Param(value=1.0, units='number')),
  Destination(name='main output', volume=Param(value=0.0, units='decibels'), mute=False),
  0,
  0)]
```

I also didn't have to convert the output of Multiply back into the [-1, 1] range. The `LFO` suggestion makes sense, though -- will look into that and update if it works better!
